### PR TITLE
[Otel] Refactor Boto3 instrumentation

### PR DESCRIPTION
**Pull request by GSVarsha**
coverage: 100%

### CODE REVIEW

File: src/instana/instrumentation/boto3_inst.py [hunk 1]
```diff
@@ -5,142 +5,182 @@
...
+    from botocore.client import BaseClient
 
 try:
-    import opentracing as ot
     import boto3
     from boto3.s3 import inject
 
-    def extract_custom_headers(span, headers):
+    def extract_custom_headers(span: "InstanaSpan", headers: Dict[str, Any]) -> None:
         if agent.options.extra_http_headers is None or headers is None:
```
Comment (pvital) [inline src/instana/instrumentation/boto3_inst.py:27]: If the `headers` can be `None`, we should update the type hint to `Optional[Dict[str, Any]] = None` and also update the if:
```suggestion
    def extract_custom_headers(span: "InstanaSpan", headers: Optional[Dict[str, Any]] = None) -> None:
        if not agent.options.extra_http_headers or not headers:
```

File: tests/clients/boto3/test_boto3_lambda.py [hunk 2]
```diff
@@ -1,95 +1,96 @@
...
 
-        self.assertEqual(boto_span.t, test_span.t)
-        self.assertEqual(boto_span.p, test_span.s)
+        assert boto_span.t == test_span.t
+        assert boto_span.p == test_span.s
 
-        self.assertIsNone(test_span.ec)
-        self.assertIsNone(boto_span.ec)
+        assert test_span.ec is None
+        assert boto_span.ec is None
```
Comment (pvital) [inline tests/clients/boto3/test_boto3_lambda.py:54]:
```suggestion
        assert not test_span.ec
        assert not boto_span.ec
```
You can adapt the rest of the tests with this change.

File: tests/clients/boto3/test_boto3_lambda.py [hunk 3]
```diff
@@ -1,95 +1,96 @@
...
-        self.assertEqual(boto_span.data['http']['url'], f'{endpoint}:443/Invoke')
-
-    def test_lambda_invoke_as_root_exit_span(self):
+        assert boto_span.data['boto3']['ep'] == endpoint
+        assert boto_span.data['boto3']['reg'] == self.lambda_region
+        assert 'FunctionName' in boto_span.data['boto3']['payload']
+        assert boto_span.data['boto3']['payload']['FunctionName'] == self.function_name
+        assert boto_span.data['http']['status'] == 200
+        assert boto_span.data['http']['method'] == 'POST'
+        assert boto_span.data['http']['url'] == f'{endpoint}:443/Invoke'
```
Comment (pvital) [inline tests/clients/boto3/test_boto3_lambda.py:64]: This whole block is using single quotes on strings. Please, update to double quotes in the rest of the file and the other test files.

```
assert boto_span.data["boto3"]["op"] == "Invoke"
```

File: tests/clients/boto3/test_boto3_lambda.py [hunk 4]
```diff
@@ -1,95 +1,96 @@
...
+        assert result["StatusCode"] == 200
         result_payload = json.loads(result["Payload"].read().decode("utf-8"))
-        self.assertIn("message", result_payload)
-        self.assertEqual("success", result_payload["message"])
+        assert "message" in result_payload
+        assert "success" == result_payload["message"]
 
         spans = self.recorder.queued_spans()
-        self.assertEqual(1, len(spans))
+        assert 1 == len(spans)
```
Comment (pvital) [inline tests/clients/boto3/test_boto3_lambda.py:76]:
```suggestion
        assert len(spans) == 1
```

File: tests/clients/boto3/test_boto3_lambda.py [hunk 5]
```diff
@@ -1,95 +1,96 @@
...
         agent.options.allow_exit_as_root = True
         result = self.aws_lambda.invoke(FunctionName=self.function_name, Payload=json.dumps({"message": "success"}))
 
-        self.assertEqual(result["StatusCode"], 200)
+        assert result["StatusCode"] == 200
         result_payload = json.loads(result["Payload"].read().decode("utf-8"))
-        self.assertIn("message", result_payload)
-        self.assertEqual("success", result_payload["message"])
+        assert "message" in result_payload
+        assert "success" == result_payload["message"]
```
Comment (pvital) [inline tests/clients/boto3/test_boto3_lambda.py:73]:
```suggestion
        assert result_payload["message"] == "success"
```

**Review by pvital — CHANGES_REQUESTED** [review]
Few requests. Some of them should be applied multiple times, also in the other files.

File: src/instana/instrumentation/boto3_inst.py (hunk 1 above)
Comment (GSVarsha) [reply src/instana/instrumentation/boto3_inst.py]: Done.

File: tests/clients/boto3/test_boto3_lambda.py (hunk 2 above)
Comment (GSVarsha) [reply tests/clients/boto3/test_boto3_lambda.py]: Done.

File: tests/clients/boto3/test_boto3_lambda.py (hunk 3 above)
Comment (GSVarsha) [reply tests/clients/boto3/test_boto3_lambda.py]: Done.

File: tests/clients/boto3/test_boto3_lambda.py (hunk 4 above)
Comment (GSVarsha) [reply tests/clients/boto3/test_boto3_lambda.py]: Done.

File: tests/clients/boto3/test_boto3_lambda.py (hunk 5 above)
Comment (GSVarsha) [reply tests/clients/boto3/test_boto3_lambda.py]: Done.

**Review by pvital — APPROVED** [review]
It looks good to me.

**Pull request merged**